### PR TITLE
added coercion support for Debezium time types via annotations

### PR DESCRIPTION
**Pull request by waffletower-reify**
#### Description

added coercion support for Debezium time types via annotations.
Provides "human-readable" local-date and microsecond timestamp coercions from fixed-point integer types to formatted strings.
Behavior is switchable via configuration flags, and is disabled by default for backwards compatibility.

##### PR Checklist
- [x] Tests added
- [x] [Changelog](CHANGELOG.md) updated 

[DE-1488]

[DE-1488]: https://reifyhealth.atlassian.net/browse/DE-1488?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 1]
```diff
@@ -128,7 +150,10 @@ private Type.PrimitiveType icebergFieldType(String fieldType) {
 
   private Object jsonValToIcebergVal(Types.NestedField field,
                                      JsonNode node) {
-    LOGGER.debug("Processing Field:{} Type:{}", field.name(), field.type());
+    String fieldTypeName = field.doc();
+    LOGGER.debug("Processing Field:{} Type:{} Doc:{}",
+                field.name(), field.type(), fieldTypeName);
```
Comment (dacreify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java]: Nit :)
```suggestion
    LOGGER.debug("Processing Field:{} Type:{} Doc:{}",
                 field.name(), field.type(), fieldTypeName);
```

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 2]
```diff
@@ -147,8 +172,20 @@ private Object jsonValToIcebergVal(Types.NestedField field,
...
+        else {
+          if (IcebergChangeEvent.coerceDebeziumMicroTimestamp &&
+              fieldTypeName.equals("io.debezium.time.MicroTimestamp")) {
+            val = node.isNull() ? null : Instant.ofEpochSecond(0L, node.asLong() * 1000).toString();
+          }
+          else {
+            // if the node is not a value node (method isValueNode returns false), convert it to string.
+            val = node.isValueNode() ? node.asText(null) : node.toString();
+          }
+        }
```
Comment (dacreify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java]: Style idea:

```suggestion
        // string destination coercions based upon schema 'name' annotations
        if (IcebergChangeEvent.coerceDebeziumDate &&
            fieldTypeName.equals("io.debezium.time.Date")) {
          val = node.isNull() ? null
                : LocalDate.ofEpochDay(node.asInt()).toString();
        }
        else if (IcebergChangeEvent.coerceDebeziumMicroTimestamp &&
                 fieldTypeName.equals("io.debezium.time.MicroTimestamp")) {
          val = node.isNull() ? null
                : Instant.ofEpochSecond(0L, node.asLong() * 1000).toString();
        }
        else {
          // if the node is not a value node (method isValueNode returns false), convert it to string.
          val = node.isValueNode() ? node.asText(null)
                : node.toString();
        }
```

File: src/test/java/com/getindata/kafka/connect/iceberg/sink/TestIcebergUtil.java [hunk 3]
```diff
@@ -131,6 +131,73 @@ public void valuePayloadWithSchemaAsJsonNode() {
...
+        Schema schema = e.icebergSchema();
+        GenericRecord record = e.asIcebergRecord(schema);
+        String schemaString = schema.toString();
+        String recordString = record.toString();
+
+        assertTrue(schemaString.contains("ship_date: optional int (io.debezium.time.Date)"));
+        assertTrue(schemaString.contains("ship_timestamp: optional long (io.debezium.time.MicroTimestamp)"));
+        assertTrue(recordString.contains("77663"));
+        assertTrue(recordString.contains("6710075456016196"));
+    }
```
Comment (dacreify) [inline src/test/java/com/getindata/kafka/connect/iceberg/sink/TestIcebergUtil.java]: Nice coverage of all three states of the configuration: Defaulted, explicit disable, explicit enable 👍

**Review by dacreify — APPROVED** [review]
Looks great! Just a couple style ideas

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java (hunk 2 above)
Comment (waffletower-reify) [reply src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java]: Good idea!

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 4]
```diff
@@ -97,14 +108,25 @@ private GenericRecord asIcebergRecord(Types.StructType tableFields, JsonNode dat
...
 
-  private Type.PrimitiveType icebergFieldType(String fieldType) {
+  private Type.PrimitiveType icebergFieldType(String fieldType, String fieldTypeName) {
     switch (fieldType) {
       case "int8":
       case "int16":
       case "int32": // int 4 bytes
-        return Types.IntegerType.get();
+        if (IcebergChangeEvent.coerceDebeziumDate && fieldTypeName.equals("io.debezium.time.Date")) {
+          return Types.StringType.get();
```
Comment (dacreify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java]: Date primitives are supported in Iceberg -- any reason we wouldn't use that here?

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 5]
```diff
@@ -97,14 +108,25 @@ private GenericRecord asIcebergRecord(Types.StructType tableFields, JsonNode dat
...
+        }
+        else {
+          return Types.IntegerType.get();
+        }
       case "int64": // long 8 bytes
-        return Types.LongType.get();
+        if (IcebergChangeEvent.coerceDebeziumMicroTimestamp &&
+            fieldTypeName.equals("io.debezium.time.MicroTimestamp")) {
+          return Types.StringType.get();
```
Comment (dacreify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java]: Timestamps are also supported -- can we use those?

https://iceberg.apache.org/spec/

Comment (waffletower-reify) [reply src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java]: I'll look into those, will probably have to still coerce the timestamp.  But the Date might be supported from epochdays

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 6]
```diff
@@ -304,19 +345,27 @@ private List<Types.NestedField> icebergSchema(JsonNode eventSchema, String schem
...
+                                                         Types.StructType.of(subSchema), ""));
             columnId += subSchema.size();
             break;
           default: //primitive types
-            schemaColumns.add(Types.NestedField.optional(columnId, fieldName, icebergFieldType(fieldType)));
+            // passing fieldTypeName for NestedField `doc` attribute,
+            // annotation based value coercions can be made utilizing the NestedField `doc` initializer/method
+            schemaColumns.add(Types.NestedField.optional(columnId, fieldName,
+                                                         icebergFieldType(fieldType, fieldTypeName),
+                                                         fieldTypeName));
```
Comment (dacreify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java:390]: @waffletower-reify Are these related to the date stuff?

Comment (dacreify) [reply src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java:390]: Oh I see, it's to pass through the Debezium type

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 7]
```diff
@@ -97,14 +97,30 @@ private GenericRecord asIcebergRecord(Types.StructType tableFields, JsonNode dat
...
   }
 
-  private Type.PrimitiveType icebergFieldType(String fieldType) {
+  private Type.PrimitiveType icebergFieldType(String fieldType, String fieldTypeName) {
     switch (fieldType) {
       case "int8":
       case "int16":
       case "int32": // int 4 bytes
-        return Types.IntegerType.get();
+        if (configuration.isRichTemporalTypes() &&
```
Comment (waffletower-reify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java:105]: good having the single configuration for all types

File: src/test/java/com/getindata/kafka/connect/iceberg/sink/TestIcebergUtil.java [hunk 8]
```diff
@@ -131,6 +138,110 @@ public void valuePayloadWithSchemaAsJsonNode() {
...
+                .withCustomProperty("rich-temporal-types", "true")
+                .build();
+        IcebergChangeEvent e = new IcebergChangeEvent(
+                "test",
+                MAPPER.readTree(debeziumTimeCoercionSchema).get("payload"), null,
+                MAPPER.readTree(debeziumTimeCoercionSchema).get("schema"), null,
+                configuration
+        );
+        Schema schema = e.icebergSchema();
+
```
Comment (waffletower-reify) [inline src/test/java/com/getindata/kafka/connect/iceberg/sink/TestIcebergUtil.java:216]: Nice,  good type coverage

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java [hunk 9]
```diff
@@ -35,15 +33,17 @@ public class IcebergChangeEvent {
...
   private final JsonNode key;
   private final JsonSchema jsonSchema;
+  private final IcebergSinkConfiguration configuration;
 
   public IcebergChangeEvent(String destination,
                             JsonNode value,
                             JsonNode key,
                             JsonNode valueSchema,
-                            JsonNode keySchema) {
+                            JsonNode keySchema, IcebergSinkConfiguration configuration) {
```
Comment (waffletower-reify) [inline src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java:42]: I went with what I thought was a more performant solution by injecting class variable state rather than passing the configuration object, as there will be a lot of iteration on the configuration if it is passed in wholesale.  This is definitely cleaner, but adds extra hash table lookups per field processed which impacts performance a bit, particularly considering that the switch change does not change during up time.   But if I had done this in clojure, I may have passed the configuration everywhere as a map as well initially.

**Review by waffletower-reify — COMMENTED** [review]
Look good, I can't approve though, as I am a contributor.

File: src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java (hunk 9 above)
Comment (dacreify) [reply src/main/java/com/getindata/kafka/connect/iceberg/sink/IcebergChangeEvent.java:42]: Yeah I don't think this will create a performance bottleneck. The previous solution was drawn to my attention because it caused a unit test to regress due to the mutation of global state.

Which hash table are you referring to? I thought it was just object reference passing and access, which as you say would be replete in a Clojure solution.
